File: Gissenger/SkillSweep.py
import matplotlib.pyplot as plt
import numpy as np

from mpl_toolkits import mplot3d
from PredictionFunctions import *

def skillSweep(xMatrix,reversal_times,dt):
    #Initialize threshold vector
    threshold_vector = np.array([[],[],[]])
    #Number of steps minus 2
    sweep_parameter = 50
    #Sweep of thresholds
    for i in range(1,sweep_parameter):
        for j in range(1,sweep_parameter):
            sweep = runPrediction(xMatrix,i/sweep_parameter,j/sweep_parameter,reversal_times,dt)
            mcc = sweep[3][0]
            if mcc == 0:
                continue
            current_threshold = np.array([[i/sweep_parameter],[j/sweep_parameter],[mcc]])
            threshold_vector = np.append(current_threshold,threshold_vector,axis=1)

    #Indicate optimal threshold
    optimal_threshold_pos = threshold_vector[0][np.where(threshold_vector[2][:] == np.amax(threshold_vector[2][:]))]
    optimal_threshold_neg = threshold_vector[1][np.where(threshold_vector[2][:] == np.amax(threshold_vector[2][:]))]
    print("The optimal positive threshold(s) is/are: ",optimal_threshold_pos,", and the optimal negative threshold(s) is/are: ",optimal_threshold_neg," with an MCC of: ",np.amax(threshold_vector[2][:]))
    #Plot Threshold values vs MCC
    ax = plt.axes(projection = '3d')
    ax.set(xlabel='Positive Threshold',ylabel='Negative Threshold',zlabel='MCC')
    ax.plot3D(threshold_vector[0][:],threshold_vector[1][:],threshold_vector[2][:], 'go', label='MCC vs Thresholds')
    plt.show()

File: GissengerMatlab/skillSweep.m
function skillSweep(xMatrix,reversal_times,dt)

    threshold_vector = double.empty(3,0);
    sweep_parameter = 50;

    for i = 1:sweep_parameter
        for j = 1:sweep_parameter
            sweep = runPrediction(xMatrix,i/sweep_parameter,j/sweep_parameter,reversal_times,dt);
            slabel = label(xMatrix,i/sweep_parameter,j/sweep_parameter);
            mcc = sweep(4,1);
            if mcc == 0
                continue;
            end
            current_threshold = [i/sweep_parameter;j/sweep_parameter;mcc];
            threshold_vector = [threshold_vector current_threshold];
        end
    end
    %Find column index where max MCC occurs
    [row,column] = find(threshold_vector(3,:) == max(threshold_vector(3,:)));
    %Indicate optimal threshold
    optimal_threshold_pos = threshold_vector(1,column);
    optimal_threshold_neg = threshold_vector(2,column);
    fprintf("The optimal positive threshold(s) is/are: %i",optimal_threshold_pos);
    fprintf(", and the optimal negative threshold(s) is/are: %i",optimal_threshold_neg);
    fprintf(" with an MCC of: %i", max(threshold_vector(3,:)));
    %Plot Threshold values vs MCC
    figure
    for kk = 1:length(threshold_vector)
        if mod(kk,10)==0
            plot3(threshold_vector(1,1:kk),threshold_vector(2,1:kk),threshold_vector(3,1:kk),'-','LineWidth',1)
            hold on,plot3(threshold_vector(1,kk),threshold_vector(2,kk),threshold_vector(3,kk),'.','Color','r','MarkerSize',20)
            axis([-2.5 2.5 -2.5 2.5 -2.5 2.5])
            view([90 0])
            drawnow
            hold off
        end
    end
    
